india_compliance/gst_india/doctype/purchase_reconciliation_tool/purchase_reconciliation_tool.py

```diff
@@ -0,0 +1,694 @@
+
+        if "Previous Month" in period:
+            start_month = end_month = month - 1 or 12
+        elif "Quarter" in period:
+            end_month = quarter * 3
+            if "Previous" in period:
+                end_month = (quarter - 1) * 3 or 12
+            start_month = end_month - 2
+
+        if start_month > month:
+            start_year = end_year = year - 1
+        elif "Year" in period:
+            start_month = 4
+            end_month = 3
+            if "Previous" in period:
+                start_year = end_year = year - 1
+            if start_month > month:
+                start_year -= 1
+            else:
+                end_year += 1
+
+        date1 = _get_first_day(start_month, start_year)
+        date2 = _get_last_day(end_month, end_year)
+        date2 = datetime.strftime(date2 if date2 < now else now, "%Y-%m-%d")
+        return [date1, date2]
+
+    @frappe.whitelist()
+    def get_reconciliation_data(self):
+        purchase = self.query_purchase_invoice()
+        inward_supply = self.query_inward_supply()
+
+        # this will not return missing in purchase (if any)
+        summary_data = (
+            purchase.left_join(inward_supply)
+            .on(
+                (inward_supply.link_doctype == "Purchase Invoice")
+                & (inward_supply.link_name == self.pi.name)
+            )
+            .select(
+                *self.get_inward_supply_fields(for_summary=True, table=inward_supply)
+            )
+            .where(
+                (self.pi.posting_date[self.purchase_from_date : self.purchase_to_date])
+                | (
+                    (inward_supply.link_doctype == "Purchase Invoice")
+                    & (inward_supply.link_name == self.pi.name)
+                )
+            )
+            .run(as_dict=True)
+        )
+
+        # add missing in purchase data
+        summary_data = summary_data + (
+            self.query_inward_supply(for_summary=True)
+            .where(
+                (self.gstr2.link_doctype != "Purchase Invoice")
+                | (self.gstr2.link_name == "")
+                | (self.gstr2.link_name.isnull())
+            )
+            .run(as_dict=True)
+        )
+
+        return summary_data
+
+
+def get_periods(fiscal_year, return_type: ReturnType):
+    """Returns a list of month (formatted as `MMYYYY`) in a fiscal year"""
+
+    fiscal_year = frappe.db.get_value(
+        "Fiscal Year",
+        fiscal_year,
+        ("year_start_date as start_date", "year_end_date as end_date"),
+        as_dict=True,
+    )
+
+    if not fiscal_year:
+        return []
+
+    end_date = min(fiscal_year.end_date, _getdate(return_type))
+
+    # latest to oldest
+    return tuple(reversed(_get_periods(fiscal_year.start_date, end_date)))
+
+
+def _get_periods(start_date, end_date):
+    """Returns a list of month (formatted as `MMYYYY`) in given date range"""
+
+    return pd.date_range(start_date, end_date, freq="MS").strftime("%m%Y").tolist()
```

Comment:
We need to remove `pandas` as a dependency (reasons: [1](https://github.com/frappe/erpnext/pull/30598#issuecomment-1090978603) [2](https://github.com/frappe/erpnext/issues/27047)). It is an overkill for this use-case.

Reply:
I think this answer can achieve same result:
https://stackoverflow.com/a/28290050/4767738

```
import datetime
from dateutil.rrule import rrule, MONTHLY

strt_dt = datetime.date(2001,1,1)
end_dt = datetime.date(2005,6,1)

dates = [dt for dt in rrule(MONTHLY, dtstart=strt_dt, until=end_dt)]
```